rasterio/_base.pyx

```diff
@@ -108,11 +108,24 @@ def driver_can_create_copy(drivername):
 cdef class DatasetBase(object):
     """Dataset base class."""
 
-    def __init__(self, path, options=None):
+    def __init__(self, path=None, options=None):
```

Comment:
@sgillies What if instead of accepting a path/connection string, `DatasetBase()` accepted an open dataset handle?  This could be useful for exposing some GDAL features that are responsible for their own I/O.  For example, as mentioned in https://github.com/mapbox/rasterio/pull/1041#issuecomment-308625948, [`GDALCreateCopy()`](http://www.gdal.org/gdal_8h.html#a249cb5d462af0784d44754b3668059be) returns an open dataset.  Another issue I encountered but didn't mention is that we don't have a generic place to put that dataset handle.  A lot of the generic dataset properties are just derived from a dataset handle and could be implemented as properties rather than being forced to explicitly set each one in `start()` (now `__init__()`).  We could just introduce a `DatasetReader|DatasetWriter.from_path()` to handle the `GDALOpen()` calls.  I suspect this could simplify the `WarpedVRT()` stack a bit too.

For instance:

```
from rasterio._io import DatasetReader

def copy(src, dst, driver):
    hds = GDALCreateCopy(src, dst)
    if GDALAccess(hds) == GA_ReadOnly:
        return DatasetReader(hds)
    else:
        return DatasetWriter(hds)


with rasterio.copy(src, dst) as ds:
    print(ds.mode)
```

Or for `WarpedVRT()`:

```
class WarpedVRTReaderBase(DatasetReaderBase):

    def __init__(src, src_crs, dst_crs, ...):
        # Most of the current __init__
        hds = GDALCreateWarpedVRT(...)
        super(WarpedVRTReaderBase, self).__init__(hds)
```